Inconsistency in PreTrainedModel.resize_token_embeddings When ZeRO3 Is Enabled

sinamoeini opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

This PR addresses https://github.com/huggingface/transformers/issues/25241.

In previous implementation when ZeRO stage 3 was enbaled, resize_token_embeddings would create independent PyTorch weights on each device. Here we ensure that new embeddings are created with DeepSpeed init, and are properly partitioned accros devices.



## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

- @pacman100
- @ydshieh

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @sgugger

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam

Documentation: @sgugger, @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: @sgugger
- TensorFlow: @Rocketknight1

 -->

sinamoeini-amz committed:
Inconsistency in PreTrainedModel.resize_token_embeddings

This PR addresses https://github.com/huggingface/transformers/issues/25241.

In previous implementation when ZeRO stage 3 was enbaled, resize_token_embeddings would create independent PyTorch weights on each device. Here we ensure that new embeddings are created with DeepSpeed init, and are properly partitioned accros devices.

sinamoeini-amz committed:
formatting with black

ydshieh commented on src/transformers/modeling_utils.py:
```diff
@@ -1464,24 +1464,34 @@ def _get_resized_embeddings(
-            with deepspeed.zero.GatheredParameters(old_embeddings.weight, modifier_rank=0):
-                if torch.distributed.get_rank() == 0:
-                    new_embeddings.weight.data[:n, :] = old_embeddings.weight.data[:n, :]
+            with deepspeed.zero.Init(config_dict_or_path=deepspeed_config()):
+                new_embeddings = nn.Embedding(
+                    new_num_tokens,
+                    old_embedding_dim,
+                    device=old_embeddings.weight.device,
+                    dtype=old_embeddings.weight.dtype,
+                )
+
+            params = [old_embeddings.weight, new_embeddings.weight]
+            with deepspeed.zero.GatheredParameters(params, modifier_rank=0):
+                self._init_weights(new_embeddings)
+                new_embeddings.weight.data[:n, :] = old_embeddings.weight.data[:n, :]
```
would like @pacman100 and @sgugger to take a look here.

ydshieh commented on src/transformers/modeling_utils.py:
```diff
@@ -1464,24 +1464,34 @@ def _get_resized_embeddings(
             import deepspeed
 
-            with deepspeed.zero.GatheredParameters(old_embeddings.weight, modifier_rank=0):
-                if torch.distributed.get_rank() == 0:
```
I am not familiar on this topic, but why this line is removed? I am sure there is good reason.

sinamoeini replied:
I think github diff is a bit misleading here. These two lines were replaced with (see above)
```
            params = [old_embeddings.weight, new_embeddings.weight]
            with deepspeed.zero.GatheredParameters(params, modifier_rank=0):
```
The only thing  I removed was 
```
 if torch.distributed.get_rank() == 0:
```
since it is a parallel process it is not saving latency. I can add it back in if you think it would make more sense.

ydshieh reviewed (commented):
Overall LGTM, but it's better for @pacman100 and the core maintainer to take a look, especially for the DeepSpeed part of code.

Thank you for your work!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_25394). All of your documentation changes will be reflected on that endpoint.

shady273185Gunit commented on src/transformers/modeling_utils.py:
```diff
@@ -1464,24 +1464,34 @@ def _get_resized_embeddings(
-        # initialize all new embeddings (in particular added tokens)
-        self._init_weights(new_embeddings)
-
-        # Copy token embeddings from the previous weights
```
Ek smaak n kus koop en jou daar in sit

sinamoeini replied:
Will add them back in, after other reviewers give me their feedbacks

ydshieh replied:
Before merge, don't forget this comment 🙏

sgugger reviewed (approved):
This looks good to me, but like the expert @pacman100 to chime in too.
Thanks for your PR!

sinamoeini commented:
@pacman100 would you mind reviewing this PR?

pacman100 reviewed (approved):
Thank you @sinamoeini for the fixes, LGTM! 🤗

sinamoeini commented:
Thank you @sgugger would you mind merging this?

ydshieh commented:
There is a comment to be addressed please 🙏

sinamoeini-amz committed:
adding the removed comments back in

sinamoeini commented:
@ydshieh I addressed your comment however the documentation build cancelled. Could you kick it off again and merge if there are no more issues? thanks

ydshieh commented:
It's fine. Thanks again 🚀 !
